std/algorithm/iteration.d

```diff
@@ -4289,3 +4304,625 @@ unittest
+    repeat = number of repetitions
+
+Returns:
+    Forward range which yields the product items
+*/
+auto cartesianPower(Range)(Range r, size_t repeat=1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in
+{
+    assert(repeat >= 1, "Invalid repeat");
+}
+body
+{
+    import std.array: array;
+    import std.range: indexed;
+    static struct CartesianPower
+    {
+        private ElementType!Range[] _front;
+        private Range _r;
+        private size_t[] _state;
+
+        private size_t _repeat, _nrElements;
+        private size_t _max_states, _pos;
+        private bool _isNrStatesComputed;
+
+        private bool _empty;
+
+        this(Range r, size_t repeat) pure nothrow @safe
+        {
+            _r = r;
+            // cache number of elements
+            _nrElements = r.length;
+            _repeat = repeat;
+
+            // set initial state and calculate max possibilities
+            _max_states = length;
+            if (_nrElements > 0 && _max_states > 0)
+            {
+                _state.length = repeat;
+                _front = _r.indexed(_state).array;
+            }
+            else
+                _empty = true;
+        }
+
+        @property typeof(this) save()
+        {
+            // we need to save more information about the current state
+            auto c = typeof(this)(_r, _repeat);
+            c._state = _state.dup;
+            c._pos = _pos;
+            c._empty = _empty;
+            return c;
+        }
+
+        @property ElementType!Range[] front() pure nothrow @nogc
+        {
+            return _front;
+        }
+
+        void popFront() pure nothrow @safe
+        {
+            if (empty) return;
+            // check whether we have reached the end
+            _pos++;
+            if (_pos == _max_states)
+            {
+                _empty = true;
+                return;
+            }
+            /*
+            * Bitwise increment - starting from back
+            * It works like adding 1 in primary school arithmetic.
+            * If a block has reached the number of elements, we reset it to
+            * 0, and continue to increment, e.g. for n = 2:
+            *
+            * [0, 0, 0] -> [0, 0, 1]
+            * [0, 1, 1] -> [1, 0, 0]
+            */
+            foreach_reverse (i, ref el; _state)
+            {
+                ++el;
+                if (el < _nrElements)
+                    break;
+
+                el = 0;
+            }
+            _front = _r.indexed(_state).array;
+        }
+
+        @property size_t length() pure nothrow @nogc
+        {
+            if (!_isNrStatesComputed){
+                if (_nrElements > 0)
+                {
+                    import std.math: pow;
+                    _max_states = pow(_nrElements, _repeat);
+                }
+                _isNrStatesComputed = true;
+            }
+            return _max_states;
+        }
+
+        @property bool empty() pure nothrow @safe
```

Comment:
Doesn't need attribute inference, but please either use it or specify all attributes explicitly, i.e. it's missing `@nogc`